### PR TITLE
added main file for pred visualization

### PR DESCRIPTION
**Pull request by VictorGillioz**
PR for visualising a model prediction on a random mesh from a validation dataset, can be in or out of distribution

plot

### CODE REVIEW

File: meshnets/visualize.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import torch
+from torch.utils.data import random_split
+
+from meshnets.utils.datasets import FromDiskGeometricDataset
+from meshnets.utils import data_visualization
+from meshnets.utils import model_loading
+
+FLAGS = flags.FLAGS
+
+# Random seed flag
```
Comment (AugustoPeres) [inline meshnets/visualize.py:18]: Remove these comments

File: meshnets/visualize.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    sample_idx = np.random.choice(validation_dataset.indices)
+
+    mesh_path = dataset.get_mesh_path(sample_idx)
+    graph_path = dataset.get_graph_path(sample_idx)
+
+    graph = torch.load(graph_path)
+    graph = wrapper.normalize_labels(graph)
+    normalized_groundtruth = graph.y.detach().numpy()
+
+    prediction = wrapper(graph).detach().numpy()
```
Comment (AugustoPeres) [inline meshnets/visualize.py:70]: I think that you may want to call `wrapper.eval()` to tell pytorch_lightning that this is to eval the model. This good practice as it will shutdown dropouts, use inference batch norm, .... Even if such layer do not exist, this is good practice regardless

When using the trainer.eval(model, dataloader) it is not required as lightning will do it under the hood. But here I think it is.

Comment (AugustoPeres) [reply meshnets/visualize.py:70]: Also, instead of `wrapper(graph).detach()` use `with torch.no_grad()`. This will prevent tensor being put in the computational graph, saving memory and eliminating the need of then removing them from there

**Review by AugustoPeres — APPROVED** [review]
LGTM, left just a few comments

File: meshnets/utils/data_visualization.py [hunk 3]
```diff
@@ -24,6 +23,19 @@ def plot_mesh(mesh_path: str,
...
+
+    if scalars is None:
+        raise TypeError("Argument 'scalars' is None.")
+
+    if scalars.ndim == 1:
+        return scalars
+    elif scalars.ndim == 2 and scalars.shape[1] == 1:
+        return scalars.flatten()
+    elif scalars.ndim > 1:
+        raise ValueError("Argument 'scalars' must have a single dimension.")
```
Comment (fabiocruz) [inline meshnets/utils/data_visualization.py:36]: I think this should be an `else` clause (or even better, the clause should be omitted). The message itself is also not accurate, since by line 33 `scalars.ndim` can be 2. Rephrase the error message.

File: meshnets/visualize.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+flags.DEFINE_string('tracking_uri', None,
+                    'The tracking URI for the MLFlow experiments.')
+
+flags.DEFINE_string('run_id', None, 'The run id of the experiment to load.')
+
+flags.DEFINE_integer('checkpoint', 0,
+                     'The checkpoint to load from the experiment.')
+
+flags.DEFINE_bool('start_xvfb', False, 'Wheter to start xvfb or not.')
```
Comment (fabiocruz) [inline meshnets/visualize.py:37]: Typo, "whether"

File: meshnets/visualize.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+flags.DEFINE_string('save_dir', 'img', 'Directory where to save the plots.')
+
+flags.DEFINE_string('file_name', 'plot.png', 'The name of the the plot file.')
```
Comment (fabiocruz) [inline meshnets/visualize.py:41]: Why not have one flag only with `output_file_path`?

File: meshnets/visualize.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+    data_visualization.plot_mesh_comparison(mesh_path,
+                                            normalized_groundtruth,
+                                            prediction,
+                                            clim=None,
+                                            rot_z=180,
+                                            off_screen=False,
+                                            screenshot=os.path.join(
+                                                FLAGS.save_dir,
+                                                FLAGS.file_name))
```
Comment (fabiocruz) [inline meshnets/visualize.py:81]: No reason for these two flags to exist. One would be enough.

File: meshnets/visualize.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                                                   FLAGS.checkpoint)
+
+    sample_idx = np.random.choice(validation_dataset.indices)
+
+    mesh_path = dataset.get_mesh_path(sample_idx)
+    graph_path = dataset.get_graph_path(sample_idx)
+
+    with torch.no_grad():
+        graph = torch.load(graph_path)
+        graph = wrapper.normalize_labels(graph)
```
Comment (fabiocruz) [inline meshnets/visualize.py:65]: Is `normalize_labels` a staticmethod of the wrapper class? It should be stored elsewhere, if so.

File: meshnets/visualize.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        normalized_groundtruth = graph.y
+
```
Comment (fabiocruz) [inline meshnets/visualize.py:67]: Can you provide more context to what's happening in each of these lines (in comments to the PR, to start with).

**Review by fabiocruz — APPROVED** [review]
LGTM. I have minor(ish) comments only.

File: meshnets/visualize.py (hunk 7 above)
Comment (VictorGillioz) [reply meshnets/visualize.py:65]: It is not a static method, as the normalisation depends on the label statistics the wrapper has been initialised with.

File: meshnets/visualize.py (hunk 8 above)
Comment (VictorGillioz) [reply meshnets/visualize.py:67]: We want to compare the groundtruth, obtained from the raw pressure field and normalized according to dataset statistics, and the prediction, infered from the graph representation by the model trained with normalized labels.

File: meshnets/visualize.py (hunk 7 above)
Comment (fabiocruz) [reply meshnets/visualize.py:64]: I see, thanls.

File: meshnets/visualize.py (hunk 8 above)
Comment (fabiocruz) [reply meshnets/visualize.py:66]: Thanks.

File: meshnets/utils/data_visualization.py (hunk 3 above)
Comment (VictorGillioz) [reply meshnets/utils/data_visualization.py]: You're right, I changed it to an else clause and changed the error message. Why do you think the clause should be omitted? I wrote this method because pyvista allows faulty plots when the provided array does not conform to their required dimensions. This leads to plots that looks plausible but do not represent the correct data.
